File: server/pieces/BishopPiece.cpp
#include "BishopPiece.h"
#include "../Game.h"

#include <stdexcept>

std::list<Position> BishopPiece::getValidMoves()
{
    Piece* piece;
    std::list<Position> valid;
    int tempX, tempY;
    
    for(int x = -1; x <= 1; x+=2)
    {
        for(int y = -1; y <= 1; y+=2)
        {
            tempX = x;
            tempY = y;
            do
            {
                try
                {
                    piece = _game->getCell( Position(_myPos.x+tempX, _myPos.y+tempY) ); 
                    
                    if( piece == nullptr || (piece!=nullptr && piece->isWhite() != _white))
                        valid.push_front( Position(_myPos.x+tempX, _myPos.y+tempY) );
                    
                    tempX+=x;
                    tempY+=y;
                }
                catch(std::out_of_range &e) 
                {
                    break;
                }
            }
            while(piece == nullptr);
        }
    }
    return valid;
}

File: client/js/Pieces/BishopPiece.js
class BishopPiece extends Piece {
    constructor(id,white,pos,g,forward) {
        if(pos!=undefined) {
            var actualPos = pos.clone();
            super(id,white,actualPos,g,forward);
        }
        else {
            super(id,white,pos,g,forward);
        }
        this.visual = new BishopPieceVisual(this);

    }

    getValidMoves() {
        let valid = [];
        let tempX;
        let tempY;
        let pieceInArea = false;

        for( let x = -1; x <= 1; x+=2 ) {
            for( let y = -1; y <= 1; y+=2) {
                pieceInArea = false;
                tempX = x;
                tempY = y;
                while(pieceInArea == false) {
                    try {
                        pieceInArea = this.pushIfAvailable(valid, new Position(
                            this.pos.x + tempX, this.pos.y + tempY), undefined);
                            
                        tempX+=x;
                        tempY+=y;
                    } catch( err ) { break; }
                } 
            }
        }

        console.log("Bis");
        return valid;
    }

    setPos(pos) {
        super.setPos(pos);
    }
    
    update(deltaTime) {
        this.visual.update(deltaTime);
    }

    die() {
        super.die();
        this.visual.die();
    }
}